front/temporal/upsert_queue/activities.ts

```diff
@@ -146,3 +174,107 @@ export async function upsertDocumentActivity(
     upsertContext: upsertQueueItem.upsertContext || undefined,
   });
 }
+
+async function upsertTableItem(
+  upsertQueueItem: t.TypeOf<typeof EnqueueUpsertTableFromCsv>,
+  upsertQueueId: string,
+  enqueueTimestamp: number
+) {
+  const logger = mainLogger.child({
+    upsertQueueId,
+    workspaceId: upsertQueueItem.workspaceId,
+    dataSourceName: upsertQueueItem.dataSourceName,
+    tableId: upsertQueueItem.tableId,
+  });
+
+  const auth = await Authenticator.internalBuilderForWorkspace(
+    upsertQueueItem.workspaceId
+  );
+
+  const owner = auth.workspace();
+  if (!owner) {
+    logger.error(
+      {
+        delaySinceEnqueueMs: Date.now() - enqueueTimestamp,
+      },
+      "[UpsertQueue] Giving up: Workspace not found"
+    );
+    return;
+  }
+
+  const dataSource = await getDataSource(auth, upsertQueueItem.dataSourceName);
+
+  if (!dataSource) {
+    // If the data source was not found, we simply give up and remove the item from the queue as it
+    // means that the data source was deleted.
+    logger.info(
+      {
+        delaySinceEnqueueMs: Date.now() - enqueueTimestamp,
+      },
+      "[UpsertQueue] Giving up: DataSource not found"
+    );
+    return;
+  }
+
+  const statsDTags = [
+    `data_source_name:${dataSource.name}`,
+    `workspace_id:${upsertQueueItem.workspaceId}`,
+  ];
+
+  const upsertTimestamp = Date.now();
+
+  const tableRes = await upsertTableFromCsv({
+    owner,
+    projectId: dataSource.dustAPIProjectId,
+    dataSourceName: dataSource.name,
+    tableName: upsertQueueItem.tableName,
+    tableDescription: upsertQueueItem.tableDescription,
+    tableId: upsertQueueItem.tableId,
+    csv: upsertQueueItem.csv,
+    truncate: upsertQueueItem.truncate,
+  });
+
+  if (tableRes.isErr()) {
+    logger.error(
+      {
+        error: tableRes.error,
+        latencyMs: Date.now() - upsertTimestamp,
+        delaySinceEnqueueMs: Date.now() - enqueueTimestamp,
+      },
+      "[UpsertQueue] Failed upsert"
+    );
+    statsDClient.increment("upsert_queue_error.count", 1, statsDTags);
+    statsDClient.distribution(
+      "upsert_queue_upsert_error.duration.distribution",
```

Comment:
we want distinguished metrics here document/table

Reply:
Ack, added `update metrics & dashboard in datadog` to deploy plan